### PR TITLE
Factor out functionality from update() that will be needed for layering

### PR DESCRIPTION
**Pull request by mkenigs**
Various refactors that will be helpful for sharing functionality between a legacy and layering update flow. Opening this against `master` per discussion in https://github.com/openshift/machine-config-operator/pull/2982#issuecomment-1059193703. I would like to get a CI run because I'm having trouble launching a cluster with code we're trying to get into `layering`

### CODE REVIEW

File: pkg/daemon/update.go [hunk 1]
```diff
@@ -114,9 +114,6 @@ func reloadService(name string) error {
 }
 
 // performPostConfigChangeAction takes action based on what postConfigChangeAction has been asked.
-// For non-reboot action, it applies configuration, updates node's config and state.
-// In the end uncordon node to schedule workload.
-// If at any point an error occurs, we reboot the node so that node has correct configuration.
```
Comment (mkenigs) [inline pkg/daemon/update.go:119]: Are these comments out of date? Or am I misunderstanding how error handling works here? It looks like it just returns errors

File: pkg/daemon/drain.go [hunk 2]
```diff
@@ -153,15 +152,17 @@ func (dn *Daemon) performDrain() error {
 	return nil
 }
 
+type ReadFileFunc func(string) ([]byte, error)
```
Comment (cgwalters) [inline pkg/daemon/drain.go:155]: This looks totally fine to me as is, but I'm thinking we may want to make this an interface in the future?

File: pkg/controller/common/constants.go [hunk 3]
```diff
@@ -1,6 +1,8 @@
 package common
 
 const (
+	MCONamespace = "openshift-machine-config-operator"
```
Comment (cgwalters) [inline pkg/controller/common/constants.go:4]: Probably worth converting at least one user in this commit.

**Review by cgwalters — APPROVED** [review]
Overall looks sane to me; I am a bit less confident in the later changes just because of the complexity of the code.  But, I think a lot of this is covered by our e2es.

File: pkg/daemon/drain.go (hunk 2 above)
Comment (mkenigs) [reply pkg/daemon/drain.go:155]: Agreed. https://issues.redhat.com/browse/MCO-224

File: pkg/daemon/drain_test.go [hunk 4]
```diff
@@ -359,9 +361,15 @@ location = "example.com/repo/test-img"
 				t.Errorf("parsing new Ignition config failed: %v", err)
 			}
 			diffFileSet := ctrlcommon.CalculateConfigFileDiffs(&oldIgnConfig, &newIgnConfig)
-			drain, err := isDrainRequired(test.actions, diffFileSet, oldIgnConfig, newIgnConfig)
-			if !reflect.DeepEqual(test.expectedAction, drain) {
-				t.Errorf("Failed determining drain behavior: expected: %v but result is: %v. Error: %v", test.expectedAction, drain, err)
+			readOldFile := func(path string) ([]byte, error) {
```
Comment (cheesesashimi) [inline pkg/daemon/drain_test.go:364]: I noticed you duplicated these in a few places. A technique to reduce duplication is to have a function that returns a function (I know it sounds strange):

```golang
func getIgnitionFileDataReadFunc(ignConfig *ign3types.IgnConfig, path string) ReadFileFunc {
  return func(path string) ([]byte, error) {
    return ctrlcommon.GetIgnitionFileDataByPath(ignConfig, path)
  }
}
```

Then your calls become something like:

```golang
readOldFile := getIgnitionFileDataReadFunc(&oldIgnConfig, path)
readNewFile := getIgnitionFileDataReadFunc(&newIgnConfig, path)
```

File: pkg/daemon/update.go [hunk 5]
```diff
@@ -428,50 +428,58 @@ func (dn *Daemon) applyOSChanges(mcDiff machineConfigDiff, oldConfig, newConfig
...
+		}
+		glog.Infof("Setting post config change action to postConfigChangeActionReboot; %s present", constants.MachineConfigDaemonForceFile)
+		return []string{postConfigChangeActionReboot}, nil
+	}
+
 	filesPostConfigChangeActionNone := []string{
 		"/etc/kubernetes/kubelet-ca.crt",
 		"/var/lib/kubelet/config.json",
 	}
+	if dn.os.IsFCOS() {
```
[the remark this answers is not in the document]
Comment (cheesesashimi) [reply pkg/daemon/update.go:447]: I was thinking about this as well.

**Review by cheesesashimi — COMMENTED** [review]
I just have a couple minor questions and suggestions.

File: pkg/daemon/update.go [hunk 6]
```diff
@@ -625,7 +622,16 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig) (retErr err
...
 	}
 
-	return dn.performPostConfigChangeAction(actions, newConfig.GetName())
+	if err := dn.performPostConfigChangeAction(actions, newConfig.GetName()); err != nil {
+		return err
+	}
+
+	// if dn.skipReboot, a reboot might be performed manually after update(), in which case we don't want to update state
+	// else reboot was not needed to apply the configuration, and we need to update state
+	if !dn.skipReboot {
```
Comment (yuqi-zhang) [inline pkg/daemon/update.go:631]: I'm a bit confused at the use of skipReboot here. This is only for onceFrom right? Why is this specifically handled here?

**Review by yuqi-zhang — COMMENTED** [review]
Some minor comments. I think it would be helpful for me at least if I can understand why some of these refactors are needed, for example:

1. Why does calculatePostConfigChangeAction need to be consolidated? 
2. What is the extra FCOS checks added for?
3. Why do we need to move isDrainRequired -> drainIfRequired

As a counter example, the explanation around readFileFunc is very nice

File: pkg/daemon/drain.go [hunk 7]
```diff
@@ -153,26 +152,40 @@ func (dn *Daemon) performDrain() error {
...
-// isDrainRequired determines whether node drain is required or not to apply config changes.
-func isDrainRequired(actions, diffFileSet []string, oldIgnConfig, newIgnConfig ign3types.Config) (bool, error) {
+type ReadFileFunc func(string) ([]byte, error)
+
+// drainIfRequired determines whether node drain is required or not to apply config changes and performs the drain if required
+// returns whether drain was attempted
+func (dn *Daemon) drainIfRequired(actions, diffFileSet []string, readOldFile, readNewFile ReadFileFunc) (bool, error) {
+	// For any unhandled cases (although currently there aren't any), default to drain
+	drain := true
 	if ctrlcommon.InSlice(postConfigChangeActionReboot, actions) {
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply pkg/daemon/drain.go:162]: I don't think it's actually redundant because technically actions is a list (although currently it's only ever of length 1). So if we didn't have this if branch, if actions included both postConfigChangeActionReboot and postConfigChangeActionReloadCrio we could take eg the postConfigChangeActionReloadCrio branch and skip a drain

File: pkg/daemon/update.go (hunk 6 above)
Comment (mkenigs) [reply pkg/daemon/update.go:631]: I missed adding this condition originally and it was causing bootstrap to fail. Looking at the old version of `performPostConfigChangeAction` it calls `dn.reboot` which I initially assumed would never return. But if `dn.skipReboot == true`, `dn.reboot` does return, and `performPostConfigChangeAction` returns early before updating state. So I had to add this to preserve that behavior. I can't say I fully understand why that's necessary because how state is managed is pretty unclear to me

File: pkg/daemon/drain.go [hunk 8]
```diff
@@ -153,26 +152,40 @@ func (dn *Daemon) performDrain() error {
 	return nil
 }
 
-// isDrainRequired determines whether node drain is required or not to apply config changes.
-func isDrainRequired(actions, diffFileSet []string, oldIgnConfig, newIgnConfig ign3types.Config) (bool, error) {
+type ReadFileFunc func(string) ([]byte, error)
+
+// drainIfRequired determines whether node drain is required or not to apply config changes and performs the drain if required
+// returns whether drain was attempted
+func (dn *Daemon) drainIfRequired(actions, diffFileSet []string, readOldFile, readNewFile ReadFileFunc) (bool, error) {
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply pkg/daemon/drain.go:159]: Yep I can change it to that way. Is the advantage just less churn?

[the remark this answers is not in the document]
Comment (cheesesashimi) [reply pkg/daemon/drain.go:159]: Not just less churn but it also provides good separation of concerns. `isDrainRequired` answers the question of whether a drain is required or not. `drainIfRequired` can use that to ask the question and then perform the drain if it's needed.

File: pkg/daemon/drain_test.go [hunk 9]
```diff
@@ -348,6 +348,8 @@ location = "example.com/repo/test-img"
 		},
 	}
 
+	// kubeClient must be nil so that performDrain is a no-op
+	dn := Daemon{}
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply pkg/daemon/drain_test.go]: Removed because of one of @cheesesashimi 's suggestions

**Comment by mkenigs** [comment]
>    1. Why does calculatePostConfigChangeAction need to be consolidated?

Added a much more detailed commit message (https://github.com/openshift/machine-config-operator/pull/2987/commits/c4b5dcf6143a895b9a278e366136303e59f77e83) Does that clear it up?

>     2. What is the extra FCOS checks added for?

The path for ssh keys is different on FCOS. Some of the explanation I added in the commit message should give some more context

>     3. Why do we need to move isDrainRequired -> drainIfRequired

Per @cheesesashimi 's feedback I just wrote a wrapper function. I'm just anticipating de-duplicating between legacy and layered update.

File: pkg/daemon/drain_test.go (hunk 4 above)
Comment (mkenigs) [reply pkg/daemon/drain_test.go]: Ooo I like that more! Changed

File: pkg/daemon/drain.go [hunk 10]
```diff
@@ -176,6 +176,20 @@ func isDrainRequired(actions, diffFileSet []string, readOldFile, readNewFile Rea
 	return true, nil
 }
 
+func (dn *Daemon) drainIfRequired(actions, diffFileSet []string, readOldFile, readNewFile ReadFileFunc) error {
```
Comment (kikisdeliveryservice) [inline pkg/daemon/drain.go:179]: Can you explain what this is doing? Because I don't quite understand or know if this refactor should be pulled into master without accompanying code to make sense of it?

File: pkg/daemon/update.go [hunk 11]
```diff
@@ -486,6 +483,14 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig) (retErr err
 			}
 		}
 	}
+	return nil
+}
```
Comment (kikisdeliveryservice) [inline pkg/daemon/update.go:495]: similar comment to here - i dont really understand why this is necessary or how it fits into things?

File: cmd/machine-config-operator/main.go [hunk 12]
```diff
@@ -8,10 +8,7 @@ import (
 	"k8s.io/component-base/cli"
 )
 
-const (
-	componentName      = "machine-config"
-	componentNamespace = "openshift-machine-config-operator"
-)
+const componentName = "machine-config"
```
Comment (kikisdeliveryservice) [inline cmd/machine-config-operator/main.go:11]: I think the namespace constant changes could just be their own PR (8 file changes) and merge asap.

File: pkg/daemon/drain.go (hunk 10 above)
Comment (mkenigs) [reply pkg/daemon/drain.go:179]: It's just factoring code out of `update()` so it doesn't have to be duplicated when we want to perform drains for a layered update

Comment (kikisdeliveryservice) [reply pkg/daemon/drain.go:179]: I can't help but feel like that's should probably live with the other code that needs it? Feels weird to merge it if it's needed for something that isn't in master  or a clear optimization for master

**Comment by mkenigs** [comment]
Per @kikisdeliveryservice comments have opened:
https://github.com/openshift/machine-config-operator/pull/3020
https://github.com/openshift/machine-config-operator/pull/3021
https://github.com/openshift/machine-config-operator/pull/3022

**Comment by mkenigs** [comment]
Will wait for other PRs to merge before revisiting commits left here

**Comment by mkenigs** [comment]
@kikisdeliveryservice is this small enough now or should I split it again?

File: pkg/daemon/drain.go (hunk 10 above)
Comment (cgwalters) [reply pkg/daemon/drain.go:179]: The counterbalance here is avoiding a later merge of layering into master being too large and painful, as well as the ongoing cost of rebases.

It's highly likely that there will be an intersection between layering and hypershift work for example.  And having prep patches merged into master helps avoid conflicts there too.

In this particular case, I think

- The code isn't *significantly* better or anything, but IMO it is a bit better
- The code isn't difficult to review
- CI covers all the functionality here

IMO, that on balance argues for a master merge.

**Comment by cgwalters** [comment]
Since this is targeting layering, I don't see any blockers to merge.
If there's stuff to address we can always do followups.
/lgtm
